src/dmd/dtoh.d

```diff
@@ -1778,39 +1802,44 @@ public:
                 first = false;
             else
                 buf.writestring(", ");
-            buf.writestring("typename ");
-            buf.writestring(p.ident.toChars());
-        }
-        buf.writestringln(">");
 
-        // TODO replace this block with a sd.accept
-        {
-            buf.writestring(sd.isUnionDeclaration() ? "union " : "struct ");
-            buf.writestring(sd.ident.toChars());
-            if (sd.members)
+            if (p.isTemplateTypeParameter())
             {
-                buf.writenl();
-                buf.writestringln("{");
-                auto savex = adparent;
-                adparent = sd;
```

Comment:
@MoonlightSentinel it's because you removed this line that the functions aren't written out as virtual.

Reply:
Though looking at the old code, it only set this for templated structs/unions, not classes.

Besides, the condition for whether or not to emit `virtual` shouldn't depend on whether `adparent` is set.  `fd.isVirtual() && fd.vtblIndex != -1` is how other parts of the compiler deal with it.  (`adparent` is used for the checkbuf assertion however).

Reply:
I see, thanks!
Not sure why `dtoh` even checks `adparent` instead of using `fd.isVirtual()`/ `fd.isVirtualFunction()`

Reply:
> Besides, the condition for whether or not to emit virtual shouldn't depend on whether adparent is set.

Ideally not, but `vtblIndex` is still `-1` for templated `ClassDeclaration`s.